repository/src/lib.rs

```diff
@@ -768,11 +772,92 @@ impl<T: RawRepository> DistributedRepository<T> {
         Ok((block_header, result))
     }
 
-    /// Creates an agenda commit on top of the `work` branch.
+    /// Creates an extra-agenda transaction commit on top of the `work` branch.
     pub async fn create_extra_agenda_transaction(
         &mut self,
-        _transaction: &ExtraAgendaTransaction,
+        transaction: &ExtraAgendaTransaction,
     ) -> Result<CommitHash, Error> {
-        unimplemented!()
+        let work_commit = self.raw.locate_branch(WORK_BRANCH_NAME.into()).await?;
+        let last_header_commit = self.raw.locate_branch(FINALIZED_BRANCH_NAME.into()).await?;
+        let reserved_state = self.get_reserved_state().await?;
+
+        // Check if the `work` branch is rebased on top of the `finalized` branch.
+        if self
+            .raw
+            .find_merge_base(last_header_commit, work_commit)
+            .await?
+            != last_header_commit
+        {
+            return Err(eyre!(
+                "branch {} should be rebased on {}",
+                WORK_BRANCH_NAME,
+                FINALIZED_BRANCH_NAME
+            ));
+        }
+
+        // Construct a commit list starting from the next commit of the last finalized block to the `branch_commit`(the most recent commit of the branch)
+        let ancestor_commits = self.raw.list_ancestors(work_commit, Some(256)).await?;
+        let position = ancestor_commits
+            .iter()
+            .position(|c| *c == last_header_commit)
+            .expect("TODO: handle the case where it exceeds the limit.");
+        let commits = stream::iter(ancestor_commits.iter().take(position).rev().cloned().map(
+            |c| {
+                let raw = &self.raw;
+                async move { raw.read_semantic_commit(c).await.map(|x| (x, c)) }
+            },
+        ))
+        .buffered(256)
+        .collect::<Vec<_>>()
+        .await;
+        let mut commits = commits.into_iter().collect::<Result<Vec<_>, _>>()?;
+        // Add most recent commit of the branch to the list since it is not included in the ancestor commits
```

Comment:
```suggestion
        // Add the most recent commit of the branch to the list since it is not included in the ancestor commits
```

Reply:
Fixed. But this is actually an outdated code not using the function in utils. See [26d7009](https://github.com/postech-dao/simperby/pull/178/commits/26d7009bb9ed882d61ccdb0138a4df8ec73c34e8)